### PR TITLE
fix(dropdown): support ReactNode type value

### LINKED ISSUE #50135: Dropdown children不能传入ReactNode类型，跟实际类型不符
### Reproduction link

[Edit on StackBlitz](https://stackblitz.com/edit/react-wayfrm?file=demo.tsx)

### Steps to reproduce

打开分享链接

### What is expected?

兼容 React.ReactNode类型

### What is actually happening?

React.Children.only expected to receive a single React element child.

| Environment | Info |
| --- | --- |
| antd | 5.19.3 |
| React | 18.3.3 |
| System | Mac |
| Browser | 谷歌 |

### PR DESCRIPTION
**Pull request by coding-ice**
[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [x] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
close: https://github.com/ant-design/ant-design/issues/50135

### 📝 Changelog

- Use a developer-oriented tone and narrative style.
- Describe the user's first-hand experience of the issue and its impact on developers, rather than your solution approach.
- Refer to: https://ant.design/changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | support ReactNode type value          |
| 🇨🇳 Chinese | 支持 ReactNode 类型的值          |

### CODE REVIEW

File: components/dropdown/dropdown-button.tsx [hunk 1]
```diff
@@ -13,6 +13,8 @@ import type { DropdownProps } from './dropdown';
 
 export type DropdownButtonType = 'default' | 'primary' | 'dashed' | 'link' | 'text';
 
+type ButtonsRenderType = [React.ReactNode, React.ReactElement];
```
Comment (crazyair) [inline components/dropdown/dropdown-button.tsx:16]: 不应该都是 ReactNode 吗

Comment (coding-ice) [reply components/dropdown/dropdown-button.tsx:16]: 如果是`string`之类的，代码直接崩掉了
`const child = React.Children.only(children) as React.ReactElement<any>`

Comment (coding-ice) [reply components/dropdown/dropdown-button.tsx:16]: 看起来处理起来，并是不是那么容易，这样一处理其他的地方都挂了

Comment (crazyair) [reply components/dropdown/dropdown-button.tsx:16]: 组件里可以判断 isElement 吧？

Comment (coding-ice) [reply components/dropdown/dropdown-button.tsx:16]: 还是兼容下代码，不能去改类型

Comment (crazyair) [reply components/dropdown/dropdown-button.tsx:16]: 对，不改类型，改代码兼容

File: components/dropdown/dropdown.tsx [hunk 2]
```diff
@@ -175,7 +175,13 @@ const Dropdown: CompoundedComponent = (props) => {
 
   const [, token] = useToken();
 
-  const child = React.Children.only(children) as React.ReactElement<any>;
+  const child = React.useMemo(() => {
```
Comment (li-jia-nan) [inline components/dropdown/dropdown.tsx:178]: 这里没必要用 useMemo

Comment (li-jia-nan) [reply components/dropdown/dropdown.tsx:178]: 一个三元就搞定了

File: components/dropdown/dropdown.tsx [hunk 3]
```diff
@@ -176,7 +176,9 @@ const Dropdown: CompoundedComponent = (props) => {
 
   const [, token] = useToken();
 
-  const child = React.Children.only(children) as React.ReactElement<any>;
+  const child = React.Children.only(
+    React.isValidElement(children) ? children : <span>{children}</span>,
```
Comment (zombieJ) [inline components/dropdown/dropdown.tsx:180]: 改 warning 吧，这样包元素是个 break change。如果用户用 tooltip 之类的会叠加事件传递的都会乱掉

Comment (coding-ice) [reply components/dropdown/dropdown.tsx]: 我把简单数据类型都处理了下，麻烦再看下～

File: components/_util/isPrimitive.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+const isPrimitive = (value: unknown) => (typeof value !== 'object' && typeof value !== 'function') || value === null;
```
Comment (li-jia-nan) [inline components/_util/isPrimitive.ts:1]: 这样的话把 `undefined` 和 `null` 也包含进去了，我的理解是只需要 `typeof value === 'string' || typeof value === 'number';` 就可以了

Comment (coding-ice) [reply components/_util/isPrimitive.ts:1]: `null`和`undefined`应该也是要处理的，同属于`ReactNode`类型，传入也崩掉了

**Comment by afc163** [comment]
冲突了

**Review by afc163 — DISMISSED** [review]

**Review by afc163 — APPROVED** [review]

**Pull request merged**